### PR TITLE
fix(proxy): record OpenAI cache write tokens

### PR DESCRIPTION
**Pull request by Brisbanehuang**
## Summary

- Record the explicit OpenAI `cache_write_tokens` usage field as canonical `cache_creation_input_tokens`.
- Normalize inclusive OpenAI input usage for `codex` and `openai-compatible` by subtracting both cache-read and cache-write buckets.
- Preserve canonical top-level priority, provider boundaries, and conservative behavior when the cache-write field is absent or invalid.

## Related Issues

- Follow-up to #1133 — extends the cache-subset normalization introduced there (`adjustUsageForProviderType` + `PROVIDERS_WITH_CACHE_SUBSET_USAGE`). #1133 subtracted only cache-read tokens from the inclusive input total; this PR additionally subtracts cache-write tokens and parses the previously-ignored `cache_write_tokens` field.
- Related to #1075 — the original OpenAI token double-counting report (fixed by #1133). This PR closes the residual gap: the cache-write bucket was still being dropped, leaving `cache_creation_input_tokens` unset and ordinary input inflated for OpenAI Responses/Chat Completions payloads that report `cache_write_tokens`.

## Reproduction and root cause

A real OpenAI Responses usage payload reported:

```json
{
  "input_tokens": 240951,
  "input_tokens_details": {
    "cached_tokens": 240128,
    "cache_write_tokens": 740
  },
  "output_tokens": 58
}
```

The previous parser recognized `240128` cache-read tokens but ignored the explicit `740` cache-write tokens. It therefore persisted `823` ordinary input tokens and no cache-creation bucket. The correct normalized buckets are `83` ordinary input, `740` cache write, `240128` cache read, and `58` output.

The root cause was twofold: nested OpenAI usage parsing only read `cached_tokens`, and the OpenAI subset normalization subtracted only cache-read tokens from the inclusive input total.

This change does not infer cache-write tokens from differences when `cache_write_tokens` is absent. Such inference is unsafe because a missing provider field does not prove that the remaining input-token difference represents cache creation. In that boundary case, cache creation remains unset and only an explicitly reported cache-read bucket is normalized.

## Supported shapes and boundaries

- Responses API: `input_tokens_details.cache_write_tokens`
- Chat Completions: `prompt_tokens_details.cache_write_tokens`
- Chat Completions SSE: final usage chunk using the same details shape
- Existing top-level `cache_creation_input_tokens` wins, including an explicit zero
- Negative or non-finite nested values are ignored
- Ordinary input is clamped at zero when explicit cache buckets exceed the inclusive total
- Providers outside `codex` and `openai-compatible` retain disjoint cache semantics

## Verification

- `bun run typecheck` - passed
- `bun run lint` - passed with 194 warnings, 2 infos, and 0 errors; commit `219402a7` contains the isolated OpenAPI test baseline fix
- `bun run test tests/unit/proxy/extract-usage-metrics.test.ts` - passed, 65 tests
- `LC_ALL=C bun run test` - passed, 743 files and 6731 tests; 2 files and 13 tests skipped
- `bun run build` - passed
- `git diff --check` - passed

Independent review found no Critical or Important issues and assessed the change ready to merge.

---
*Description enhanced by Claude AI*

<details open><summary><h3>Greptile Summary</h3></summary>

This PR fixes double-counting of OpenAI cache tokens by parsing the previously-ignored `cache_write_tokens` field from `input_tokens_details`/`prompt_tokens_details` and subtracting it from the inclusive `input_tokens` for `codex` and `openai-compatible` providers (following the pattern established in #1133 for `cache_read_input_tokens`).

- **`extractUsageMetrics`**: Adds a new `getNestedOpenAICacheWriteTokens` helper and a new code block that populates `cache_creation_input_tokens` from the nested OpenAI field when no top-level value already exists, respecting the established priority (top-level > nested > TTL-derived).
- **`adjustUsageForProviderType`**: Extends the input adjustment to subtract both the cached-read and cached-write buckets from the inclusive `input_tokens`, receiving the raw usage record via a new `rawUsage` parameter so it can read the same nested fields.
- **Tests**: 13 new unit tests covering Responses API, Chat Completions, SSE, top-level priority override, zero explicit value, absent field (no inference), negative/non-finite guards, and clamp-at-zero behavior.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

The change is safe to merge: it only extends usage-metric parsing to capture a previously-ignored OpenAI field, with no risk of breaking existing normalization or Anthropic/Gemini code paths.

The logic is internally consistent: getNestedOpenAICacheWriteTokens is the single source of truth for the nested field, called with the same raw object in both extractUsageMetrics and adjustUsageForProviderType, so there is no double-counting risk. Edge cases are all handled and verified by the new test suite. Non-OpenAI providers are unaffected.

No files require special attention.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/app/v1/_lib/proxy/response-handler.ts | Adds asNonNegativeFiniteNumber and getNestedOpenAICacheWriteTokens helpers, hoists detail-object references, extends extractUsageMetrics to populate cache_creation_input_tokens from nested OpenAI fields, and adds a rawUsage param to adjustUsageForProviderType to subtract write tokens from inclusive input — all with correct priority ordering and guards. |
| tests/unit/proxy/extract-usage-metrics.test.ts | Adds 13 new tests covering the reproduction case, Chat Completions, SSE, priority rules, zero explicit value, no-inference-when-absent, non-finite/negative guards, and clamp-at-zero; all cases verified correctly. |
| tests/api/api-openapi-spec.test.ts | Minor safe refactor to eliminate an unsafe optional chaining lint warning; no behavioral change. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["Raw usage payload\n(OpenAI / codex)"] --> B["extractUsageMetrics(usage)"]

    B --> C{"top-level\ncache_creation_input_tokens?"}
    C -- "yes (number)" --> D["result.cache_creation_input_tokens\n= top-level value"]
    C -- "no" --> E["getNestedOpenAICacheWriteTokens(usage)\n→ input_tokens_details.cache_write_tokens\n   or prompt_tokens_details.cache_write_tokens"]
    E -- "found > 0" --> F["result.cache_creation_input_tokens\n= nested cache_write_tokens"]
    E -- "absent / invalid" --> G["cache_creation_input_tokens\nleft undefined"]

    B --> H["Parse cache_read_input_tokens\nfrom *.cached_tokens"]
    B --> I["Parse input_tokens\nfrom input_tokens / prompt_tokens"]

    D & F & G --> J["UsageMetrics (extracted)"]
    H --> J
    I --> J

    J --> K["adjustUsageForProviderType\n(usage, providerType, rawUsage)"]

    K --> L{"codex or\nopenai-compatible?"}
    L -- "no" --> M["return usage unchanged"]
    L -- "yes" --> N{"input_tokens\nis a number?"}
    N -- "no" --> M
    N -- "yes" --> O["cachedTokens = cache_read_input_tokens ?? 0\ncacheWriteTokens = getNestedOpenAICacheWriteTokens(rawUsage) ?? 0"]
    O --> P["adjustedInput = max(inputTokens - cachedTokens - cacheWriteTokens, 0)"]
    P --> Q["return usage with input_tokens = adjustedInput"]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A["Raw usage payload\n(OpenAI / codex)"] --> B["extractUsageMetrics(usage)"]

    B --> C{"top-level\ncache_creation_input_tokens?"}
    C -- "yes (number)" --> D["result.cache_creation_input_tokens\n= top-level value"]
    C -- "no" --> E["getNestedOpenAICacheWriteTokens(usage)\n→ input_tokens_details.cache_write_tokens\n   or prompt_tokens_details.cache_write_tokens"]
    E -- "found > 0" --> F["result.cache_creation_input_tokens\n= nested cache_write_tokens"]
    E -- "absent / invalid" --> G["cache_creation_input_tokens\nleft undefined"]

    B --> H["Parse cache_read_input_tokens\nfrom *.cached_tokens"]
    B --> I["Parse input_tokens\nfrom input_tokens / prompt_tokens"]

    D & F & G --> J["UsageMetrics (extracted)"]
    H --> J
    I --> J

    J --> K["adjustUsageForProviderType\n(usage, providerType, rawUsage)"]

    K --> L{"codex or\nopenai-compatible?"}
    L -- "no" --> M["return usage unchanged"]
    L -- "yes" --> N{"input_tokens\nis a number?"}
    N -- "no" --> M
    N -- "yes" --> O["cachedTokens = cache_read_input_tokens ?? 0\ncacheWriteTokens = getNestedOpenAICacheWriteTokens(rawUsage) ?? 0"]
    O --> P["adjustedInput = max(inputTokens - cachedTokens - cacheWriteTokens, 0)"]
    P --> Q["return usage with input_tokens = adjustedInput"]
```

</a>
</details>

<sub>Reviews (2): Last reviewed commit: ["test(api): avoid unsafe optional chainin..."](https://github.com/ding113/claude-code-hub/commit/219402a79f6f76ebd52b7325906030f618d42b3e) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=43620446)</sub>

### CODE REVIEW

File: src/app/v1/_lib/proxy/response-handler.ts [hunk 1]
```diff
@@ -3854,22 +3875,26 @@ function adjustUsageForProviderType(
...
+  const hasCacheWriteTokens = typeof usage.cache_creation_input_tokens === "number";
 
-  if (typeof cachedTokens !== "number" || typeof inputTokens !== "number") {
+  if (typeof inputTokens !== "number" || (!hasCachedTokens && !hasCacheWriteTokens)) {
     return usage;
   }
 
-  const adjustedInput = Math.max(inputTokens - cachedTokens, 0);
+  const cachedTokens = hasCachedTokens ? (usage.cache_read_input_tokens ?? 0) : 0;
+  const cacheWriteTokens = hasCacheWriteTokens ? (usage.cache_creation_input_tokens ?? 0) : 0;
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply src/app/v1/_lib/proxy/response-handler.ts]: Fixed in 2e5420f5. The normalization path now validates cache-read values with asNonNegativeFiniteNumber, while nested cache-write values go through getNestedOpenAICacheWriteTokens, which applies the same finite, non-negative check. I also added a regression test covering exponent-overflow values.

File: src/app/v1/_lib/proxy/response-handler.ts [hunk 2]
```diff
@@ -3854,22 +3875,26 @@ function adjustUsageForProviderType(
...
+  const adjustedInput = Math.max(inputTokens - cachedTokens - cacheWriteTokens, 0);
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply src/app/v1/_lib/proxy/response-handler.ts]: Fixed in 2e5420f5. OpenAI subset normalization now subtracts only raw nested input_tokens_details.cache_write_tokens or prompt_tokens_details.cache_write_tokens. Explicit top-level cache_creation_input_tokens and TTL-derived cache creation remain disjoint; regression tests cover both cases.
